File: LR4/LR4/arrays.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <cctype>
#include <string>

using namespace std;

template<typename T>
void print_container(const string& prefix, const T& container) {
    cout << prefix;
    for (const auto& elem : container) {
        cout << elem << " ";
    }
    cout << endl;
}

bool is_prime(int n) {
    if (n <= 1) return false; 
    for (int i = 2; i * i <= n; ++i) {
        if (n % i == 0) return false;
    }
    return true;
}

bool has_repeated_digits(int n) {
    bool seen[10] = { false }; // Массив для отслеживания встреченных цифр
    if (n == 0) return false;
    int temp = n;
    while (temp > 0) {
        int digit = temp % 10;
        if (seen[digit]) {
            return true; // Цифра уже встречалась
        }
        seen[digit] = true;
        temp /= 10;
    }
    return false;
}

int main() {
    setlocale(LC_ALL, "Russian");

    random_device rd;
    mt19937_64 gen(rd());
    cout << "--- Задания 1, 2, 3 ---" << endl;

    const int n = 20;
    vector<int> arr1(n), arr2(n);
    uniform_int_distribution<> dist1(10, 100);

    for (int& val : arr1) val = dist1(gen);
    for (int& val : arr2) val = dist1(gen);

    print_container("Исходный массив arr1:", arr1);
    print_container("Исходный массив arr2:", arr2);
    cout << endl;

    cout << "Простые числа из диапазона [0, 10] в arr1: ";
    for (int val : arr1) {
        if (val >= 0 && val <= 10 && is_prime(val)) {
            cout << val << " ";
        }
    }
    cout << endl;

    cout << "Простые числа из диапазона [0, 10] в arr2: ";
    for (int val : arr2) {
        if (val >= 0 && val <= 10 && is_prime(val)) {
            cout << val << " ";
        }
    }
    cout << endl;
    cout << "(Числа не найдены, так как диапазон инициализации [10, 100] не содержит простых чисел из диапазона [0, 10])" << endl << endl;

    // Сортировка первой половины arr1 по возрастанию
    sort(arr1.begin(), arr1.begin() + n / 2);

    // Сортировка второй половины arr2 по убыванию
    sort(arr2.begin() + n / 2, arr2.end(), greater<int>());

    print_container("arr1 (первая половина отсортирована):", arr1);
    print_container("arr2 (вторая половина отсортирована):", arr2);
    cout << endl;

    // Обмен отсортированной части arr1 с несортированной частью arr2
    swap_ranges(arr1.begin(), arr1.begin() + n / 2, arr2.begin());

    cout << "Модификация согласно заданию:" << endl;
    print_container("Финальный arr1 (несортированные части):", arr1);
    print_container("Финальный arr2 (отсортированные части):", arr2);
    cout << "------------------------------------------" << endl << endl;


    cout << "--- Задание 4 ---" << endl;
    const int char_array_size = 25;
    string alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    uniform_int_distribution<> dist_char(0, alphabet.length() - 1);

    vector<char> char_arr(char_array_size);
    for (char& c : char_arr) {
        c = alphabet[dist_char(gen)];
    }

    print_container("Исходный массив символов:", char_arr);

    // Замена регистра
    for (char& c : char_arr) {
        if (islower(c)) {
            c = toupper(c);
        }
        else if (isupper(c)) {
            c = tolower(c);
        }
    }

    print_container("Массив с измененным регистром:", char_arr);
    cout << "------------------------------------------" << endl << endl;


    cout << "--- Задание 5 ---" << endl;
    const int int_array_size = 15;
    uniform_int_distribution<> dist2(1000, 2000);

    vector<int> source_arr(int_array_size);
    for (int& val : source_arr) {
        val = dist2(gen);
    }
    print_container("Исходный массив [1000, 2000]:", source_arr);

    // Сортировка по убыванию
    sort(source_arr.begin(), source_arr.end(), greater<int>());
    print_container("Отсортированный массив:", source_arr);

    // Создание нового массива с числами, где есть повторяющиеся цифры
    vector<int> filtered_arr;
    for (int val : source_arr) {
        if (has_repeated_digits(val)) {
            filtered_arr.push_back(val);
        }
    }

    print_container("Массив с числами, содержащими повторяющиеся цифры:", filtered_arr);
    cout << "------------------------------------------" << endl;

    return 0;
}

File: LR4/LR4/arrays.py
import random
import string

def print_container(prefix, container):
    """Функция для вывода контейнера с префиксом"""
    print(prefix, end="")
    for elem in container:
        print(elem, end=" ")
    print()

def is_prime(n):
    """Проверка, является ли число простым"""
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

def has_repeated_digits(n):
    """Проверка, есть ли в числе повторяющиеся цифры"""
    if n == 0:
        return False
    
    seen = set()
    temp = n
    while temp > 0:
        digit = temp % 10
        if digit in seen:
            return True  # Цифра уже встречалась
        seen.add(digit)
        temp //= 10
    return False

def main():
    print("--- Задания 1, 2, 3 ---")
    
    # Инициализация генератора случайных чисел
    random.seed()
    
    n = 20
    arr1 = [random.randint(10, 100) for _ in range(n)]
    arr2 = [random.randint(10, 100) for _ in range(n)]
    
    print_container("Исходный массив arr1:", arr1)
    print_container("Исходный массив arr2:", arr2)
    print()
    
    # Поиск простых чисел в диапазоне [0, 10]
    print("Простые числа из диапазона [0, 10] в arr1: ", end="")
    primes_arr1 = [val for val in arr1 if 0 <= val <= 10 and is_prime(val)]
    for val in primes_arr1:
        print(val, end=" ")
    print()
    
    print("Простые числа из диапазона [0, 10] в arr2: ", end="")
    primes_arr2 = [val for val in arr2 if 0 <= val <= 10 and is_prime(val)]
    for val in primes_arr2:
        print(val, end=" ")
    print()
    print("(Числа не найдены, так как диапазон инициализации [10, 100] не содержит простых чисел из диапазона [0, 10])")
    print()
    
    # Сортировка первой половины arr1 по возрастанию
    arr1[:n//2] = sorted(arr1[:n//2])
    
    # Сортировка второй половины arr2 по убыванию
    arr2[n//2:] = sorted(arr2[n//2:], reverse=True)
    
    print_container("arr1 (первая половина отсортирована):", arr1)
    print_container("arr2 (вторая половина отсортирована):", arr2)
    print()
    
    # Обмен отсортированной части arr1 с несортированной частью arr2
    arr1[:n//2], arr2[:n//2] = arr2[:n//2], arr1[:n//2]
    
    print("Модификация согласно заданию:")
    print_container("Финальный arr1 (несортированные части):", arr1)
    print_container("Финальный arr2 (отсортированные части):", arr2)
    print("------------------------------------------")
    print()
    
    print("--- Задание 4 ---")
    char_array_size = 25
    alphabet = string.ascii_letters  # 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
    
    char_arr = [random.choice(alphabet) for _ in range(char_array_size)]
    
    print_container("Исходный массив символов:", char_arr)
    
    # Замена регистра
    for i in range(len(char_arr)):
        if char_arr[i].islower():
            char_arr[i] = char_arr[i].upper()
        elif char_arr[i].isupper():
            char_arr[i] = char_arr[i].lower()
    
    print_container("Массив с измененным регистром:", char_arr)
    print("------------------------------------------")
    print()
    
    print("--- Задание 5 ---")
    int_array_size = 15
    
    source_arr = [random.randint(1000, 2000) for _ in range(int_array_size)]
    print_container("Исходный массив [1000, 2000]:", source_arr)
    
    # Сортировка по убыванию
    source_arr.sort(reverse=True)
    print_container("Отсортированный массив:", source_arr)
    
    # Создание нового массива с числами, где есть повторяющиеся цифры
    filtered_arr = [val for val in source_arr if has_repeated_digits(val)]
    
    print_container("Массив с числами, содержащими повторяющиеся цифры:", filtered_arr)
    print("------------------------------------------")

if __name__ == "__main__":
    main()